### PR TITLE
REPL command completion is cursor sensitive

### PR DESCRIPTION
**Pull request by som-snytt**
Completes `:lo^x.s` to `:load^x.s` and `:lo` to `:load ^`.

Fixes scala/bug#10757
Fixes scala/bug#10468

### CODE REVIEW

File: src/repl/scala/tools/nsc/interpreter/LoopCommands.scala [hunk 1]
```diff
@@ -116,22 +117,34 @@ trait LoopCommands { self: { def echo(msg: String): Unit } =>
...
+            val suffix = cmd.name.substring(cursor - 1)   // account for colon
+            val completion = if (cmd.isInstanceOf[NullaryCmd] || cursor < line.length) suffix else suffix + " "
+            new Completion {
+              def resetVerbosity(): Unit = ()
+              def complete(buffer: String, cursor: Int) = Candidates(cursor, List(completion))
+            }
+          case cmd :: rest =>
+            new Completion {
+              def resetVerbosity(): Unit = ()
+              def complete(buffer: String, cursor: Int) = Candidates(cursor, cmds.map(_.name))
```
Comment (hrhino) [inline src/repl/scala/tools/nsc/interpreter/LoopCommands.scala]: Should this be `Candidates(cursor, cmds.map(_.name.drop(cursor - 1)))`?

As it is, `:imp` completes to `:impimp` with suggestions for `imports` and `implicits`.

Comment (som-snytt) [reply src/repl/scala/tools/nsc/interpreter/LoopCommands.scala]: I think I had that version at some point. Lemme check. (I remember :imp working.  :imp: )

**Comment by som-snytt** [comment]
In my history,

```
scala> :settings
-Xprint = List()
-Ybrowse = List()
-Ytyper-debug = false
-Ywarn-unused = UnusedWarnings.ValueSet(imports)
-Ywarn-unused-import = true
-d = .
-deprecation = true
-encoding = UTF-8
-feature = true
-nowarn = false
```

**Comment by som-snytt** [comment]
Luckily the only person who will ever exercise this feature is also reviewing it.

Tweaked and retweaked. Deleted my first usage of `Regex.Groups` as it wasn't using the `Match` to check bounds of first group.

Not sure about forward-portability for repl refactor.

**Comment by som-snytt** [comment]
I had moved the bug milestone up but not the PR, doing that in case we can test all the edge cases in one work week.

**Review by hrhino — APPROVED** [review]
Thanks. I've been recently asked to put my software development work to the side at my job to help out with QA, so I'm in a nitpicky mood, it seems. 

The fact that the repl accepts semicolon commands such as `:set -Xlog-implicits` made me think that `:set` was a command for some time when I started using it for non-trivial things. 

My hope is that someone, somewhere, eventually runs into it and that this doesn't become my bespoke feature. I had poked at it a bit a few weeks back but it turns out I didn't understand JLine one bit. Now I just don't understand it for two bits.
